protoc-gen-twirp/generator.go

```diff
@@ -644,6 +644,14 @@ func (t *twirp) generateUtils() {
 	t.P(`func (e *internalWithCause) WithMeta(key string, val string) `, t.pkgs["twirp"], `.Error { return e }`)
 	t.P()
 
+	t.P(`// badRequestError is used when the twirp server cannot unmarshal a request`)
+	t.P(`func badRequestError(cause error, msg string) `, t.pkgs["twirp"], `.Error {`)
+	t.P(`	err := `, t.pkgs["twirp"], `.NewError(`, t.pkgs["twirp"], `.InvalidArgument, msg)`)
+	t.P(`	err = err.WithMeta("twirp_invalid_request", cause.Error())`)
+	t.P(`	return err`)
+	t.P(`}`)
```

Comment:
Maybe we should use the error code `BadRoute` instead of `InvalidArgument`.

Granted, this will result in a 404 instead of a 400 error, which may be confusing on the metrics. But it fits better with the error codes philosophy that `BadRoute` is only returned by the generated code, while `NotFound`, `InvalidArgument` and `Unimplemented` are returned by the handler code. Using generated Twirp clients should never trigger this error (that's why it used to be an internal error), but `BadRoute` is expected when doing manual API calls (e.g. curl, or custom made clients).

Reply:
I agree that `invalid_argument` feels wrong, here. It's not that an argument was invalid for the application, it's that the request was just garbled. 

I think we would need a new error code, really. The gRPC error codes which inspired Twirp's use `INTERNAL` for this case. If that's the wrong code - which I'm not totally convinced of - then we should be descriptive. I don't think BadRoute is right, either; its name implies a routing issue, which this isn't.

Reply:
I agree that a new error code would be the most correct option here, though I was concerned about possible backwards-compatibility issues with old clients receiving an unrecognized error code. It seems like at least the current version of the generated go code transforms unrecognized errors into InternalErrors, so we're safe there though it would be worth checking old versions to see how this would be handled on older clients. I'm not familiar with the way the Python clients are generally used but we don't seem to provide a well-defined list of errors to Python client users so it seems up to the users to handle this case defensively.

Reply:
I've checked a few clients. Python [should be fine](https://github.com/twitchtv/twirp/blob/337e90237d72193bf7f9fa387b5b9946436b7733/protoc-gen-twirp_python/main.go#L97).
 - [github.com/twitchtv/twirp-ruby](https://github.com/twitchtv/twirp-ruby/blob/025101ed99350d2d8995b2acf4a68c86207c3426/lib/twirp/client.rb#L67-L69) will return an Internal error
 - [github.com/thechriswalker/twirp-js](https://github.com/thechriswalker/twirp-js/blob/master/src/client.js#L67-L73) will accept anything
 - [github.com/larrymyers/protoc-gen-twirp_typescript](https://github.com/thechriswalker/twirp-js/blob/master/src/client.js#L67-L73) will accept anything
 - [github.com/twirphp/twirp](https://github.com/twirphp/twirp/blob/a76d5f30be04d4f872722cddfc4639432311996a/example/client.php#L16) will accept anything